sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/implementation/MessageUtils.java

```diff
@@ -253,4 +275,75 @@ private static TransactionalState getTransactionState(ByteBuffer transactionId,
         transactionalState.setOutcome(outcome);
         return transactionalState;
     }
+
+    public static ServiceBusMessage traceMessageSpan(ServiceBusMessage serviceBusMessage,
+        Context messageContext, String hostname, String entityPath, TracerProvider tracerProvider) {
+        Optional<Object> eventContextData = messageContext.getData(SPAN_CONTEXT_KEY);
+        if (eventContextData.isPresent()) {
+            // if message has context (in case of retries), don't start a message span or add a new context
+            return serviceBusMessage;
+        } else {
+            // Starting the span makes the sampling decision (nothing is logged at this time)
+            Context newMessageContext = messageContext
+                .addData(AZ_TRACING_NAMESPACE_KEY, AZ_TRACING_NAMESPACE_VALUE)
+                .addData(ENTITY_PATH_KEY, entityPath)
+                .addData(HOST_NAME_KEY, hostname);
+            Context eventSpanContext = tracerProvider.startSpan(AZ_TRACING_SERVICE_NAME, newMessageContext,
+                ProcessKind.MESSAGE);
+            Optional<Object> eventDiagnosticIdOptional = eventSpanContext.getData(DIAGNOSTIC_ID_KEY);
+            if (eventDiagnosticIdOptional.isPresent()) {
+                serviceBusMessage.getApplicationProperties().put(DIAGNOSTIC_ID_KEY, eventDiagnosticIdOptional.get()
+                    .toString());
+                tracerProvider.endSpan(eventSpanContext, Signal.complete());
+                serviceBusMessage.addContext(SPAN_CONTEXT_KEY, eventSpanContext);
+            }
+        }
+        return serviceBusMessage;
+    }
+
+    /*
+     * Starts a new process tracing span and attaches the returned context to the EventData object for users.
+     */
+    public static Context startProcessTracingSpan(ServiceBusReceivedMessage receivedMessage,
+        String hostname, String entityPath, TracerProvider tracerProvider, ProcessKind processKind) {
+        Object diagnosticId = receivedMessage.getApplicationProperties().get(DIAGNOSTIC_ID_KEY);
+        if (diagnosticId == null || !tracerProvider.isEnabled()) {
+            return Context.NONE;
+        }
+
+        Context spanContext = tracerProvider.extractContext(diagnosticId.toString(), Context.NONE)
+            .addData(ENTITY_PATH_KEY, entityPath)
+            .addData(HOST_NAME_KEY, hostname)
+            .addData(AZ_TRACING_NAMESPACE_KEY, AZ_TRACING_NAMESPACE_VALUE);
+        spanContext = receivedMessage.getEnqueuedTime() == null
+            ? spanContext
+            : spanContext.addData(MESSAGE_ENQUEUED_TIME, receivedMessage.getEnqueuedTime().toEpochSecond());
+        return tracerProvider.startSpan(AZ_TRACING_SERVICE_NAME, spanContext, processKind);
+    }
+
+    /*
+     * Ends the process tracing span and the scope of that span.
+     */
+    public static void endProcessTracingSpan(Context processSpanContext, Signal<Void> signal,
```

Comment:
Same here. Where is this method called from? I guess this was added for the processor only but I think the async receiver should use them since the message processor is just a wrapper around the async receiver's receive method.

Reply:
We probably do same  what .Net have done for their processor.. If we just have to do tracing for each receive message, we can do in async receiver. But if there is addition tracing for processor, which .Net is doing, we should also do.

FYI-> for receive and processor would both would be kind.Consumer ? @samvaity

Reply:
Discussed with Srikanta. We only need to processor for GA and waiting for .Net to confirm with service team about tracing and then possibly add tracing to the receiver after GA. The minimum requirement is to have parity with EventHub, which has tracing for processor but not for consumer client.

Reply:
@hemanttanwar Following what we have for eventhubs, processor spans should be consumer spans.